### PR TITLE
Convert src/js/** to TypeScript

### PR DESCRIPTION
**Pull request by kubosho**
## Abstract

- Breaking change (but possible).
- This Pull Request contains some changes:
    - Build system (Added Webpack and removed Browserify)
    - Lint (Added plugins and rules for TypeScript)
- This change only does rewrite source codes with TypeScript.

## Motivation

Increase interoperability with static typing world like TypeScript or Flowtype.

### Added new build system

This introduces [Webpack](https://webpack.js.org).

Webpack does TypeScript transpile and dependency resolution.
Gulp does other builds (e.g. Sass build), test runs, dev server launch, etc.

### Convert to TypeScript

Convert *.js to *.ts simply.

## Drawback

Increase complexity of build system.

### CODE REVIEW

**Comment by kubosho** [comment]
I still have to make corrections the test files.
Then, this PR is many changes. I want to write the PR description properly.
Therefore using Draft.

**Comment by kubosho** [comment]
@MurakamiShinyu could you review this?

File: src/html/vivliostyle-viewer.ejs [hunk 1]
```diff
@@ -11,9 +11,9 @@
 <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>
 <% if (development) { %>
 <script src="../node_modules/vivliostyle/lib/vivliostyle.dev.js"></script>
-<script src="js/main-dev.js"></script>
+<script src="js/vivliostyle-ui.dev.js"></script>
 <% } else { %>
-<script src="js/main.js"></script>
+<script src="js/vivliostyle-ui.js"></script>
```
Comment (MurakamiShinyu) [inline src/html/vivliostyle-viewer.ejs]: "js/vivliostyle-ui.js" should be "js/vivliostyle-ui.min.js"

**Review by MurakamiShinyu — CHANGES_REQUESTED** [review]

File: webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                options: {
+                    configFile: tsConfigName
+                }
+            }
+        ]
+    },
+    plugins: [
+        new webpack.DefinePlugin({
+            'process.env': {
+                NODE_ENV: JSON.stringify(process.env.NODE_ENV !== "production")
```
Comment (MurakamiShinyu) [inline webpack.config.js]: `NODE_ENV: JSON.stringify(process.env.NODE_ENV !== "production")` should be
`NODE_ENV: JSON.stringify(process.env.NODE_ENV")` ?

**Comment by kubosho** [comment]
@MurakamiShinyu Could you review this?

**Review by MurakamiShinyu — APPROVED** [review]

**Comment by MurakamiShinyu** [comment]
@kubosho Thanks, LGTM!

BTW, we decided to use Prettier default (such as 2space indent) as I wrote in our slack https://vivliostyle.slack.com/archives/CNN5GPF9V/p1570589074067100.
The code format change can be done with another pull request, but also it will be good to include the change (remove `.prettierrc.js` and do `prettier --write`) in this pull request, because it makes sense to start with the new code format for the new TypeScript code. How about it?

**Comment by kubosho** [comment]
I will send Pull Request to follow up this Pull Request,

**Comment by MurakamiShinyu** [comment]
Found regressions and fixed:
- https://github.com/vivliostyle/vivliostyle-ui/commit/e4bb42224220854470529ef32982adcd32c4c22e
- https://github.com/vivliostyle/vivliostyle-ui/commit/9ee7cc9ae5046d9226a92b528f48e2ce20ec7ce5

**Pull request merged**